### ISSUE TITLE
Feature: `returns` that creating a new mock and setting behavior

### ISSUE THREAD

**Issue by TAGC** [mockk/mockk#201]
### Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [x] I am running the latest version
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed

### Expected Behavior

Given I call a function to determine the answer for a particular mock function invocation, e.g.:

```kotlin
val foo = mockk<Foo> {
    every { bar(1) } returns createResult("a")
    every { bar(2) } returns createResult("b")
}

fun createResult(value: String): Result {
    return Result(value)
}
```

If this works, then I expect that simply creating a mock inside `createResult` (which I ignore) should have no impact on behaviour i.e. the mock `Foo` instance should behave exactly the same as it did before:

```kotlin
val foo = mockk<Foo> {
    every { bar(1) } returns createResult("a")
    every { bar(2) } returns createResult("b")
}

fun createResult(value: String): Result {
    mockk<Whatever> { every { something() } just Runs }
    return Result(value)
}
```

### Current Behavior

Creating a mock instance within a function used to determine the answer for a mock function invocation (like `createResult`) *that defines at least one mock function invocation itself* (like `every { something() } ...`) causes lookups to fail.

For example, in the first snippet a call like `foo.bar(1)` would succeed and return `Result("a")`, whereas in the second case it would create an exception like:

```
Mockk failure invoking 'bar(1)': no answer found for: Foo(#1).bar(1)
```

Creating a mock in the answer with no function invocations defined does not cause this problem i.e.

```kotlin
fun createResult(value: String): Result {
    mockk<Whatever>()
    return Result(value)
}
```

Using `answers` instead of `returns` also avoids this issue (and can be used as a workaround) i.e.

```kotlin
val foo = mockk<Foo> {
    every { bar(1) } answers { createResult("a") }
    every { bar(2) } answers { createResult("b") }
}
```

#### Context

Please provide any relevant information about your setup. This is important in case the issue is not reproducible except for under certain conditions.

* MockK version: 1.18.13.kotlin13
* OS: Windows 10 (64-bit)
* Kotlin version: 1.3.11
* JDK version: 1.8.0_191
* JUnit version: 5
* Type of test: unit test

#### Stack trace

```
Mockk failure invoking 'bar(1)': no answer found for: Foo(#1).bar(1)
Exception in thread "main" io.mockk.MockKException: no answer found for: Foo(#1).bar(1)
	at io.mockk.impl.stub.MockKStub.defaultAnswer(MockKStub.kt:95)
	at io.mockk.impl.stub.MockKStub.answer(MockKStub.kt:48)
	at io.mockk.impl.recording.states.AnsweringState.call(AnsweringState.kt:16)
	at io.mockk.impl.recording.states.AnsweringStillAcceptingAnswersState.call(AnsweringStillAcceptingAnswersState.kt:16)
	at io.mockk.impl.recording.CommonCallRecorder.call(CommonCallRecorder.kt:48)
	at io.mockk.impl.stub.MockKStub.handleInvocation(MockKStub.kt:224)
	at io.mockk.impl.instantiation.JvmMockFactoryHelper$mockHandler$1.invocation(JvmMockFactoryHelper.kt:24)
	at io.mockk.proxy.jvm.advice.Interceptor.call(Interceptor.kt:20)
	at io.mockk.proxy.jvm.advice.BaseAdvice.handle(BaseAdvice.kt:42)
	at io.mockk.proxy.jvm.advice.jvm.JvmMockKProxyInterceptor.interceptNoSuper(JvmMockKProxyInterceptor.java:45)
	at experiment.Foo$ByteBuddy$B11Uh3yh.bar(Unknown Source)
	at experiment.ExperimentKt.tryBar(experiment.kt:38)
	at experiment.ExperimentKt$main$1.invokeSuspend(experiment.kt:27)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:32)
	at kotlinx.coroutines.DispatchedTask$DefaultImpls.run(Dispatched.kt:235)
	at kotlinx.coroutines.DispatchedContinuation.run(Dispatched.kt:81)
	at kotlinx.coroutines.EventLoopBase.processNextEvent(EventLoop.kt:123)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:69)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:45)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:35)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at experiment.ExperimentKt.main(experiment.kt:21)
	at experiment.ExperimentKt.main(experiment.kt)
```

#### Minimal reproducible code (the gist of this issue)

```kotlin
package experiment

import io.mockk.MockKException
import io.mockk.Runs
import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import kotlinx.coroutines.runBlocking

// True to run the code that exposes the Mockk issue
// False to run code that allows Mockk to behave as expected
const val BREAK_MOCK = true

// Define an example SUT (system under test)
data class Result(val value: String)
interface Foo {
    fun bar(number: Int): Result
}

// Create a mock instance of the SUT and make calls to it that should succeed
fun main() = runBlocking {
    val foo = mockk<Foo> {
        every { bar(1) } returns createResult("a")
        every { bar(2) } answers { createResult("b") }
    }

    foo.tryBar(1)
    foo.tryBar(2)
}

fun createResult(value: String): Result {
    MockWhateverFactory.createAnArbitraryMock()
    return Result(value)
}

fun Foo.tryBar(number: Int) {
    try {
        bar(number)
        println("Successfully invoked 'bar($number)'")
    } catch (e: MockKException) {
        println("Mockk failure invoking 'bar($number)': ${e.message}")
    }
}

// This factory is used to produce arbitrary mocked objects
object MockWhateverFactory {

    interface Whatever {
        fun something()
    }

    fun createAnArbitraryMock() = mockk<Whatever> {
        if (BREAK_MOCK) {
            every { something() } just Runs
        }
    }
}
```

**Comment by oleksiyp** [comment]
Indeed `mockk` has a global state that is reset by `returns` clause. 

That way you are not able to start another mocking statement before `returns` is executed. Actually not sure if there is an easy way to fix it. It is kind of by design limitation.

Keep this ticket open, I'll do few experiments when will have free time.

**Comment by oleksiyp** [comment]
Related #244, fixed by a774f4a and f0e0673

**Comment by oleksiyp** [comment]
Released [1.9.1](https://github.com/mockk/mockk/releases/tag/1.9.1). Please try it out and close the ticket if everything is fine. Official docs will be updated a little bit later
